client/client.go

```diff
@@ -0,0 +1,193 @@
+	commitmentAddr stdaddr.Address) (*types.FeeAddressResponse, error) {
+	var resp *types.FeeAddressResponse
+	err := c.post(ctx, "/api/v3/feeaddress", commitmentAddr, &resp, req)
+	return resp, err
+}
+
+func (c *Client) PayFee(ctx context.Context, req types.PayFeeRequest,
+	commitmentAddr stdaddr.Address) (*types.PayFeeResponse, error) {
+	var resp *types.PayFeeResponse
+	err := c.post(ctx, "/api/v3/payfee", commitmentAddr, &resp, req)
+	return resp, err
+}
+
+func (c *Client) TicketStatus(ctx context.Context, req types.TicketStatusRequest,
+	commitmentAddr stdaddr.Address) (*types.TicketStatusResponse, error) {
+	var resp *types.TicketStatusResponse
+	err := c.post(ctx, "/api/v3/ticketstatus", commitmentAddr, &resp, req)
+	return resp, err
+}
+
+func (c *Client) SetVoteChoices(ctx context.Context, req types.SetVoteChoicesRequest,
+	commitmentAddr stdaddr.Address) (*types.SetVoteChoicesResponse, error) {
+	var resp *types.SetVoteChoicesResponse
+	err := c.post(ctx, "/api/v3/setvotechoices", commitmentAddr, &resp, req)
+	return resp, err
+}
+
+func (c *Client) post(ctx context.Context, path string, addr stdaddr.Address, resp, req interface{}) error {
+	return c.do(ctx, http.MethodPost, path, addr, resp, req)
+}
+
+func (c *Client) get(ctx context.Context, path string, resp interface{}) error {
+	return c.do(ctx, http.MethodGet, path, nil, resp, nil)
+}
+
+func (c *Client) do(ctx context.Context, method, path string, addr stdaddr.Address, resp, req interface{}) error {
+	var reqBody []byte
+	var sig []byte
+	var err error
+
+	sendBody := method == http.MethodPost
+	if sendBody {
+		reqBody, err = json.Marshal(req)
+		if err != nil {
+			return fmt.Errorf("marshal request: %w", err)
+		}
+		sig, err = c.sign(ctx, string(reqBody), addr)
+		if err != nil {
+			return fmt.Errorf("sign request: %w", err)
+		}
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, method, c.URL+path, bytes.NewReader(reqBody))
+	if err != nil {
+		return fmt.Errorf("new request: %w", err)
+	}
+	if sig != nil {
+		httpReq.Header.Set("VSP-Client-Signature", base64.StdEncoding.EncodeToString(sig))
+	}
+
+	if c.log.Level() == slog.LevelTrace {
+		dump, err := httputil.DumpRequestOut(httpReq, sendBody)
+		if err == nil {
+			c.log.Tracef("Request to %s\n%s\n", c.URL, dump)
+		} else {
+			c.log.Tracef("VSP request dump failed: %v", err)
+		}
+	}
+
+	reply, err := c.Do(httpReq)
+	if err != nil {
+		return fmt.Errorf("%s %s: %w", method, httpReq.URL.String(), err)
+	}
+	defer reply.Body.Close()
+
+	if c.log.Level() == slog.LevelTrace {
+		dump, err := httputil.DumpResponse(reply, true)
+		if err == nil {
+			c.log.Tracef("Response from %s\n%s\n", c.URL, dump)
+		} else {
+			c.log.Tracef("VSP response dump failed: %v", err)
+		}
+	}
+
+	respBody, err := io.ReadAll(reply.Body)
+	if err != nil {
+		return fmt.Errorf("read response body: %w", err)
+	}
+
+	status := reply.StatusCode
+
+	if status != http.StatusOK {
```

Comment:
Is there a reason why wallet's handling of non-400 errors is different?  An issue we had run into before was where a reverse proxy was misconfigured, or vspd was not running/crashed/etc., and the 500 errors from the RP were difficult to distinguish from API errors.

Reply:
or rather, we still wanted to parse the response body if vspd returned a 400 level error.

Reply:
I'm not entirely sure what the original logic was, but afaict the errors returned by the new code will contain either the same or more information than those returned by the old code, so I think we are covered. I've added comments to explain the logic.

Reply:
I think it's probably a little more resilient to decide how to decode the response or return an error based on the HTTP status code first, rather than reading the response body first, but I think this code should work for most situations and we should still see 500 level errors in the error message as long as the reverse proxy's response body is not a JSON object.

Reply:
I don't think client should trust HTTP status for two reasons:

- Client will need to maintain its own list of statuses which it expects vspd to return, which is annoying and liable to fall out of date.
- Intermediary servers may return the same statuses as vspd returns, eg. RP might return 500 if vspd is down, vspd might return 500 if dcrd is down.

I just tried the "500 + JSON" scenario you described and I see that the code isnt working as I thought it would. I'll write some tests to make sure it works properly.

Reply:
Tests added in #363